### DISCUSSION TITLE
Need help to use ESP32 RMT module under Zephyr

### DISCUSSION THREAD

**Discussion by Alex-Stanfield** [zephyrproject-rtos/zephyr#70575] [Q&A]
Hi there, is it possible to use the RMT module ( https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/peripherals/rmt.html ) under zephyr? I'm new to zephyr but managed already to try some sample code with zephyr.

I need to output a square wave with variable frequency in order to control stepper motors (simple direction & clock lines). I need to ramp-up, flat, slow-down the speed and have a precise number of steps so I thought a HW solution (RMT) with SW supervision (via interrupts) would be fine, however I didn't find the zephyr implementation to use RMT module. Maybe I missed something? 

I'm using the esp32_devkitc_wroom board.

Thanks in advance
Alex

**Comment by rftafas** [comment]
Hi!

RMT is not supported yet. We do support LedPWM and MCPWM for such applications. If you MUST use RMT, then your options:

- DIY, You should use hal_espressif LL APIs to write a zephyr driver for RMT.
- Open a feature request and wait 
- Use another OS that supports it (NuttX, ESP-IDF)

FWIW, it is on our backlog, but not scheduled on Q2 and Q3 of 24. Main reason is that the mentioned peripherals can cover for this application.

**Comment by Alex-Stanfield** [comment]
Thanks for all responses. I’ll look into ledpwm and mcpwm  guess.

Thanks

Sds, Alex

**Comment by LuskeyNoah** [comment]
@Alex-Stanfield I started working a bit on porting the RMT peripheral into the Zephyr driver model in this issue: https://github.com/vvvvvvvvvv-LLC/t-watch-s3/issues/8 (a project of mine). Although the driver is "out-of-tree", I have structured it in such a way that it could be upstreamed to Zephyr.

If that is interesting to you, I would absolutely love some help there. Most of the structure is there, but almost none of the implementation. I didn't really see any existing driver-APIs that were a good match for the RMT peripheral, so I created a new driver-API named `rmt_tx`. If you have any suggestions for a base driver-API that maps well to the idea of transmitting arbitrary data and configuring a carrier, I'm open to suggestions.

Anyway, I'm pausing work on this for a bit to focus on some other features, but again, would love some help if it's interesting to you (or @rftafas )
